### PR TITLE
Live - Final PR for all functionality

### PR DESCRIPTION
**Pull request by shivam-tripathi**

### CODE REVIEW

**Comment by shivam-tripathi** [comment]
To help make rebasing plausible, all changes requested in the previous PRs would be handled in the final PR, else duplicate commits might creep in.

File: src/func/editor.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (C) 2018 Shivam Tripathi
```
Comment (LordSputnik) [inline src/func/editor.js]: Same for this file - @leftmostcat wrote the main bit of code in 2016

File: src/func/imports/delete-import.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+// @flow
+
+import type {Transaction} from '../types';
+
+
+export default async function deleteImport(
+	transacting: Transaction, importId: number
+) {
+	// Get the type of the import
+	const importType = await transacting.select('type')
```
Comment (LordSputnik) [inline src/func/imports/delete-import.js]: I think you need to consider the order of deletion operations here more carefully. The vote and header deletes can both happen in parallel. As soon as those are both done, the import delete can happen. As soon as the header is deleted, the data delete can happen.

So something more like:

    const headerDeletePromise = ...

    const dataDeletePromise = headerDeletePromise.then(...)

    await Promise.all([
        headerDeletePromise,
        transacting('bookbrainz.discard_votes')...
    ])

    await transacting('bookbrainz.import')...

File: src/func/imports/delete-import.js [hunk 3]
```diff
@@ -42,7 +42,8 @@ export default async function deleteImport(
 
 	// Delete the link import record
 	await transacting('bookbrainz.link_import')
-		.where('import_id', importId).del();
+		.where('import_id', importId)
+		.update('importId', null);
```
Comment (LordSputnik) [inline src/func/imports/delete-import.js]: This needs a PR in bookbrainz-sql.

File: src/func/entity-sets.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+
+export async function updateEntitySets(
+	derivedSets: ?Array<Object>, currentEntity: ?Object, entityData: Object,
+	transacting: Transaction, orm: Object
+): Promise<?Object> {
+	// If no entity sets, return null
+	if (!derivedSets) {
+		return Promise.resolve(null);
```
Comment (LordSputnik) [inline src/func/entity-sets.js]: No need for `Promise.resolve` here.

File: src/func/entity-sets.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2018 Shivam Tripathi
```
Comment (LordSputnik) [inline src/func/entity-sets.js]: Need to add @leftmostcat to the copyright header (2016).

File: src/func/create-entity.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (C) 2018 Shivam Tripathi
```
Comment (LordSputnik) [inline src/func/create-entity.js]: Need to add me and @leftmostcat to the copyright header (2016).

File: src/func/imports/misc.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+			idArr = {id};
+		}
+		catch (err) {
+			return null;
+		}
+	}
+
+	// Returning the {id} of the origin source
+	return idArr;
```
Comment (LordSputnik) [inline src/func/imports/misc.js]: `idArr` doesn't seem to have a consistent type here - in the first section, an Array is assigned, whereas the section part of the function assigns an Object to it if the Array is empty. Is that correct?

File: src/func/imports/misc.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+export async function getOriginSourceFromId(transacting, originSourceId) {
+	const [{name}] = await transacting.select('name')
```
Comment (LordSputnik) [inline src/func/imports/misc.js]: What is the array is empty?

File: src/func/imports/misc.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+		.from('bookbrainz.origin_source')
+		.where('id', originSourceId);
+
+	return name;
+}
+
+export async function getImportDetails(transacting, importId) {
+	const [details] = await transacting.select('*')
```
Comment (LordSputnik) [inline src/func/imports/misc.js]: What is the array is empty?

File: src/func/alias.js [hunk 10]
```diff
@@ -94,10 +95,12 @@ export async function updateAliasSet(
...
-export function getAliasByIds(
+export async function getAliasByIds(
 	transacting: Transaction, ids: Array<number>
 ): Promise<Object> {
-	return transacting.select('*')
+	const aliases = await transacting.select('*')
 		.from('bookbrainz.alias')
 		.whereIn('id', ids);
+	return aliases.reduce((aliasesMap, alias) =>
+		_.extend(aliasesMap, {[alias.id]: snakeToCamel(alias)}), {});
```
Comment (LordSputnik) [inline src/func/alias.js]: Why `_.extend` and not `_.assign`?

    return aliases.reduce((aliasesMap, alias) => _.assign(aliasesMap, {[alias.id]: snakeToCamel(alias)}), {});

Or with ES6:

    return aliases.reduce((aliasesMap, alias) => ({...aliasesMap, {[alias.id]: snakeToCamel(alias)}}), {});

File: src/func/imports/discard.js [hunk 11]
```diff
@@ -18,40 +18,63 @@
...
 		}
 	}
 
+	// Has the record been deleted?
+	let isDeleted = false;
+
 	// If cast vote is decisive one, delete the records
 	if (votesCast.length === DISCARD_LIMIT) {
 		await deleteImport(transacting, importId);
+		isDeleted = true;
```
Comment (LordSputnik) [inline src/func/imports/discard.js]: `return true;`

File: src/func/imports/discard.js [hunk 12]
```diff
@@ -18,40 +18,63 @@
...
 	}
 	else if (votesCast.length < DISCARD_LIMIT) {
 		// Cast vote if it's below the limit
-		await transacting.insert({editorId, importId})
+		await transacting.insert(camelToSnake({editorId, importId}))
 			.into('bookbrainz.discard_votes');
```
Comment (LordSputnik) [inline src/func/imports/discard.js]: `return false;`

File: src/func/imports/approve-import.js [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+			method: 'insert',
+			transacting
+		});
+
+	const entity = await entityModel.refresh({
+		transacting,
+		withRelated: ['defaultAlias']
+	});
+
+	await deleteImport(transacting, importId);
```
Comment (LordSputnik) [inline src/func/imports/approve-import.js]: Before deleting the import, need to set the entityId field of `link_import`.

File: src/func/imports/misc.js (hunk 7 above)
Comment (shivam-tripathi) [reply src/func/imports/misc.js]: The names I used were during the first version I wrote, hence a bit confusing. In the first case an `idArr` is either `undefined` or `{id: string}` object. In the second case, `idArr` is `{id: ?string}` depending upon success or failure of new source creation.
However, I will refactor this. `getSourceId` sounds like it gives one `id` and not `{id: ?string}`.

**Pull request merged**